src/plan_add_hashagg.c

```diff
@@ -0,0 +1,455 @@
+#include <catalog/namespace.h>
+#include <catalog/pg_type.h>
+#include <utils/guc.h>
+#include <miscadmin.h>
+#include <utils/timestamp.h>
+#include <utils/lsyscache.h>
+#include <utils/selfuncs.h>
+
+#include "compat-msvc-enter.h"
+#include <optimizer/cost.h>
+#include "compat-msvc-exit.h"
+
+#include "planner_import.h"
+#include "plan_add_hashagg.h"
+#include "extension.h"
+#include "utils.h"
+#include "guc.h"
+#include "dimension.h"
+#include "chunk_dispatch_plan.h"
+#include "planner_utils.h"
+#include "hypertable_insert.h"
+#include "constraint_aware_append.h"
+#include "compat.h"
+
+#define INVALID_ESTIMATE -1
+#define IS_VALID_ESTIMATE(est) (est >= 0)
+
+static double estimate_max_spread_expr(PlannerInfo *root, Expr *expr);
+
+/* Estimate the max spread on a time var in terms of the internal time representation.
+ * Note that this will happen on the hypertable var in most cases. Therefore this is
+ * a huge overestimate in many cases where there is a WHERE clause on time.
+ */
+static double
+estimate_max_spread_var(PlannerInfo *root, Var *var)
+{
+	VariableStatData vardata;
+	Oid			ltop;
+	Datum		max_datum,
+				min_datum;
+	int64		max,
+				min;
+	bool		valid;
+
+	examine_variable(root, (Node *) var, 0, &vardata);
+	get_sort_group_operators(var->vartype, true, false, false, &ltop, NULL, NULL, NULL);
+	valid = get_variable_range(root, &vardata, ltop, &min_datum, &max_datum);
+	ReleaseVariableStats(vardata);
+
+	if (!valid)
+		return INVALID_ESTIMATE;
+
+	max = time_value_to_internal(max_datum, var->vartype, true);
+	min = time_value_to_internal(min_datum, var->vartype, true);
+
+	if (max < 0 || min < 0)
+		return INVALID_ESTIMATE;
+
+	return ((double) max - min);
+}
+
+static double
+estimate_max_spread_opexpr(PlannerInfo *root, OpExpr *opexpr)
+{
+	char	   *name = get_opname(opexpr->opno);
+	Expr	   *left;
+	Expr	   *right;
+	Expr	   *nonconst;
+
+	if (list_length(opexpr->args) != 2 || strlen(name) != 1)
+		return INVALID_ESTIMATE;
+
+	left = linitial(opexpr->args);
+	right = lsecond(opexpr->args);
+
+	if (IsA(left, Const))
+		nonconst = right;
+	else if (IsA(right, Const))
+		nonconst = left;
+	else
+		return INVALID_ESTIMATE;
+
+	/* adding or subtracting a constant doesn't affect the range */
+	if (name[0] == '-' || name[0] == '+')
+		return estimate_max_spread_expr(root, nonconst);
+
+	return INVALID_ESTIMATE;
+}
+
+/* estimate the max spread (max(value)-min(value)) of the expr */
+static double
+estimate_max_spread_expr(PlannerInfo *root, Expr *expr)
+{
+	if (IsA(expr, Var))
+		return estimate_max_spread_var(root, (Var *) expr);
+	if (IsA(expr, OpExpr))
+		return estimate_max_spread_opexpr(root, (OpExpr *) expr);
+	return INVALID_ESTIMATE;
+}
+
+/* For time_bucket this estimate currently works by seeing how many possible
+ * buckets there will be if the data spans the entire hypertable. Note that
+ * this is an overestimate.
+ * */
+static double
+custom_group_estimate_time_bucket(PlannerInfo *root, FuncExpr *expr, double path_rows)
```

Comment:
Not so fond of hard-coding support for `time_bucket()` and `date_trunc()`. Isn't there a way we can generalize this?

Reply:
This comment still seems relevant. I figured the point of the function lookup table was to avoid having specific functions for each type. Here the time_bucket and date_trunc functions look almost identical and the only difference seems to be time interval type. Shouldn't we have one generic estimate function that then does a switch on the type of the interval argument? Or, am i missing something?

Reply:
I don't think this is a generalizable thing. I think it's a coincidence that date_trunc and time_bucket take similar arguments in similar positions. This won't always be the case. I don't want to generalize too much here.

Reply:
Even if these functions are not completely generalizable, the fact remains that they are almost identical. It should be possible to make these functions share code in a much cleaner way. The only diff I can see is how they "get" the interval. The rest is just copied boiler-plate code. Just to give one example, one could pass the `DatumGet...` and `get_interval_period...` functions as pointers to a common "worker" function.

Reply:
I refactored the code to get rid of a lot of code duplication here. Take a look now.

Reply:
Yeah, looks better.